File: android/app/src/main/java/uk/co/cerihughes/mgm/android/ui/latestevent/LatestEventEntityViewModel.kt
package uk.co.cerihughes.mgm.android.ui.latestevent

import uk.co.cerihughes.mgm.android.model.Album
import uk.co.cerihughes.mgm.android.model.AlbumType
import uk.co.cerihughes.mgm.android.model.Image
import uk.co.cerihughes.mgm.android.model.Playlist
import uk.co.cerihughes.mgm.android.ui.AlbumArtViewModel
import uk.co.cerihughes.mgm.android.ui.SpotifyURLGenerator

class LatestEventEntityViewModel(images: List<Image>, val entityType: String, val entityName: String, val entityOwner: String, val spotifyURL: String?) : AlbumArtViewModel(images) {

    companion object {
        fun createEntityViewModel(album: Album): LatestEventEntityViewModel {
            val entityType = when (album.type) {
                AlbumType.CLASSIC -> "CLASSIC ALBUM"
                AlbumType.NEW -> "NEW ALBUM"
                else -> "ALBUM"
            }

            val spotifyURL = album.spotifyId?.let {
                SpotifyURLGenerator.createSpotifyAlbumURL(it)
            }

            return LatestEventEntityViewModel(album.images, entityType, album.name, album.artist, spotifyURL)
        }

        fun createEntityViewModel(playlist: Playlist): LatestEventEntityViewModel {
            val spotifyURL = playlist.spotifyId?.let {
                SpotifyURLGenerator.createSpotifyPlaylistURL(it)
            }

            return LatestEventEntityViewModel(playlist.images, "PLAYLIST", playlist.name, playlist.owner, spotifyURL)
        }
    }
}

File: Source/UI/LatestEvent/LatestEventEntityViewModel.swift
import UIKit

protocol LatestEventEntityViewModel: AlbumArtViewModel {
    /// The album type / playlist
    var entityType: String { get }

    /// The album / playlist name
    var entityName: String { get }

    /// The artist name / playlist owner
    var entityOwner: String { get }

    /// The spotify URL to navigate to on interaction
    var spotifyURL: URL? { get }
}

final class LatestEventEntityViewModelImplementation: AlbumArtViewModelImplementation, LatestEventEntityViewModel {
    let entityType: String
    let entityName: String
    let entityOwner: String
    let spotifyURL: URL?

    init(imageLoader: ImageLoader, album: Album) {
        entityType = album.type == .classic ? "CLASSIC ALBUM" : album.type == .new ? "NEW ALBUM" : "ALBUM"
        entityName = album.name
        entityOwner = album.artist
        spotifyURL = .createSpotifyAlbumURL(albumID: album.spotifyId)

        super.init(imageLoader: imageLoader, images: album.images)
    }

    init(imageLoader: ImageLoader, playlist: Playlist) {
        entityType = "PLAYLIST"
        entityName = playlist.name
        entityOwner = playlist.owner
        spotifyURL = .createSpotifyPlaylistURL(playlistID: playlist.spotifyId)

        super.init(imageLoader: imageLoader, images: playlist.images)
    }
}
